File: VectorNotes.DomainModel/Note.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace VectorNotes.DomainModel
{
    public class Note
    {
        public int Id { get; set; }
        public string Title { get; set; }
        public string Content { get; set; }
        public int NoteCollectionId { get; set; }
        public NoteCollection? NoteCollection { get; set; }

        public IList<Tag> Tags { get; init; }

        public Note()
        {
            Title = string.Empty;
            Content = string.Empty;
            Tags = [];
        }

        public void AddTag(Tag tag)
        {
            var found = Tags.FirstOrDefault(t => t == tag);
            if (found == null)
            {
                Tags.Add(tag);
            }
        }

        public void RemoveTag(Tag tag)
        {
            var found = Tags.FirstOrDefault(t => t == tag);
            if (found != null)
            {
                Tags.Remove(found);
            }
        }
    }
}

File: vectornotes.client/src/app/model/note.ts
import { Tag } from "./tag";

export class Note {
  id?: number;
  title?: string;
  content?: string;
  noteCollectionId?: number;
  tags?: Tag[];

  constructor(source: Note | undefined = undefined) {
    if (!source) return;

    this.id = source.id;
    this.title = source.title;
    this.content = source.content;
    this.noteCollectionId = source.noteCollectionId;
    this.tags = source.tags ? [...source.tags] : [];
  }
}
